language-examples/remedial-js/9.0-functional-programming/README.md

```diff
@@ -0,0 +1,238 @@
+
+## `.map`
+
+`Array.prototype.map` returns a new array of the same length. Map accepts a single function as an argument. The function is called once for every element in the array, the return value of the function is pushed into the new array that is returned by `.map`.
+
+
+```javascript
+
+const square = (a) => a * a
+const sequence = [1, 2, 3, 4]
+const squares = sequence.map(square)
+
+// sequence remains unchanged
+sequence // [1, 2, 3, 4]
+
+// squares is a new Array
+squares // [1, 4, 9, 16]
+
+const wrappedValues = sequence.map((val) => ({ value: val }))
+
+wrappedValues // [{ value: 1 }, { value: 2 }, { value: 3 }, { value: 4 }]
+```
+
+## `.filter`
+
+`Array.prototype.filter` returns a new array of the same length or smaller. Filter expects a predicate function (a function that returns  `true` or `false`) as an argument, if the function returns a truthy value, the item is included, if the function returns a falsey value, the item is omitted.
+
+Filter 
+
+```javascript
+const sequence = [1, 2, 3, 4, 5, 6, 7]
+const odds = sequence.filter((val) => val % 2 !== 0)
+
+odds // [1, 3, 5, 7]
+
+const buddies = [{ name: 'peter', age: 18 }, { name: 'sarah', age: 22 }, { name: 'gordon', age: 31 }, { name: 'jessica', age: 14 }]
+const drinkingBuddies = buddies.filter((buddy) => buddy.age >= 19) // yay Ontario
+
+drinkingBuddies // [{ name: 'sarah', age: 22 }, { name: 'gordon', age: 31 }]
+``` 
+
+
+## `.reduce`
+
+`Array.prototype.reduce` returns a single value of any type. Reduce accepts two arguments, a function that runs once for every item in the array, and a seed value.
+
+On the first iteration, the seed value is passed as the first argument into the provided function, each successive iteration gets passed the return value from the last iteration. The return value of the final iteration becomes the return value of `.reduce`.
+
+
+```javascript
+const sum = [1, 2, 3, 4, 5].reduce((sum, nextVal) => sum + nextVal, 0)
+sum // 15
+```
+
+If the second argument (seed value) is not provided, `reduce` will use the first value of the array as the seed and start iterating on the second item.
+
+```javascript
+const largest = [12, 52, 67, 83, 32, 81, 37, 49].reduce((largest, nextVal) => nextVal > largest ? nextVal : largest)
+largest // 83
+```
+
+Reduce can be very effective for transforming data structures.
+
+```javascript
+const pokemon = [
+  { name: 'Pikachu', type: 'Electric' },
+  { name: 'Charmander', type: 'Fire' },
+  { name: 'Bulbasaur', type: 'Poison' },
+  { name: 'Vaporeon', type: 'Water' },
+  { name: 'Flareon', type: 'Fire' },
+  { name: 'Squirtle', type: 'Water' }
+]
+
+const pokemonByType = pokemon.reduce((collection, poke) => {
+ if (!Array.isArray(collection[poke.type])) collection[poke.type] = []
+
+ collection[poke.type].push(poke.name)
+ return collection
+}, {})
+
+pokemonByType  
+/* { 
+  Electric: ['Pikachu'],
+  Fire: ['Charmander', 'Flareon'],
+  Poison: ['Bulbasaur'],
+  Water: ['Vaporeon', 'Squirtle']
+  }
+*/
+```
```

Comment:
It might be nice to mention `reduceRight` also exists, though obviously not necessary.